### PR TITLE
Integration specs and fix n+1 problems Pull request

### PR DESCRIPTION
**Pull request by HERMON-1995**
### In this project, We created integration tests for all of the views used in your project. The following are the changes:

- The N+1 problem is solved when fetching all posts and their comments for a user.
- Used Capybara to write integration tests for each view in your project.
```
User index page
user show page
User post index page
Post show page
```
#### **_Note to reviewer_**
- The data is generated using the seed.rb file
- Kindly run rails db:seed && rails db:migrate for the development
- While rails db:seed RAILS_ENV=test and rails db:migrate RAILS_ENV=test to enable you fetch the data.

#### _This screenshots of your app's logs before n+1 problem fix_

Before with N+1 problem

- After n+1 problem fix

After N+1 problem fixed

### CODE REVIEW

File: spec/views/users/users_index_spec.rb [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+  it 'displays user information on show page' do
+    user = User.create(name: 'Jane', photo: 'https://example.com/jane.jpg')
+
+    visit "/users/#{user.id}"
+
+    expect(page).to have_content('Jane')
+    expect(page).to have_selector("img[src='https://example.com/jane.jpg']")
+  end
+end
```
Comment (Nemwel-Boniface) [inline spec/views/users/users_index_spec.rb]: Good job making the unit tests. However, the following tests seem to not have been included:

- [ ] I can see the number of posts each user has written.

image


- [ ] I can see the profile picture for each user.

image

File: spec/views/users/users_show_spec.rb [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+  end
+
+  it 'links to see all posts' do
+    visit user_path(user)
+
+    click_link 'See all posts'
+
+    expect(current_path.chomp('/')).to eq(user_posts_path(user))
+  end
+end
```
Comment (Nemwel-Boniface) [inline spec/views/users/users_show_spec.rb]: It seems there is no specific test for the button for showing all the posts for a user. You might confuse it with the one shat is in the test for "When I click to see all posts" but it is required to have separate tests.
- [ ] I can see a button that lets me view all of a user's posts.

image

File: spec/views/posts/posts_index_spec.rb [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+      expect(page).to have_content(user_post.likes.count)
+    end
+
+    it 'redirects me to post show page when I click on a post' do
+      link = @first_post.title
+      click_link link
+      expect(page).to have_current_path(user_post_path(@user.id, @first_post.id))
+    end
+  end
+end
```
Comment (Nemwel-Boniface) [inline spec/views/posts/posts_index_spec.rb]: The following test seems to not be included:

- [ ] I can see a section for pagination if there are more posts than fit on the view.

image

The pagination is as easy as any other test you have written. You only need to test that the button text for the pagination is tested.

**Review by Nemwel-Boniface — CHANGES_REQUESTED** [review]
## @HERMON-1995  :wave: 

### This is a manual review
# :recycle:  Request Changes :recycle: 

:heavy_check_mark: all the current tests pass :heavy_check_mark: 

image


- [ ] Kindly have a look at the comments left below. 
- [ ] KIndly ensure you include all the unit tests that are missing.




Cheers and Happy coding!

File: spec/views/users/users_index_spec.rb (hunk 1 above)
Comment (HERMON-1995) [reply spec/views/users/users_index_spec.rb]: I have added the required tests. 😊

File: spec/views/users/users_show_spec.rb (hunk 2 above)
Comment (HERMON-1995) [reply spec/views/users/users_show_spec.rb]: I have added the required tests. 😊

File: spec/views/posts/posts_index_spec.rb (hunk 3 above)
Comment (HERMON-1995) [reply spec/views/posts/posts_index_spec.rb]: I have added the required tests. 😊

File: spec/views/users/users_index_spec.rb [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+    Post.create(title: 'Post 2', text: 'Content 2', author_id: user1.id)
+    Post.create(title: 'Post 3', text: 'Content 3', author_id: user2.id)
+  end
+
+  it 'displays the list of users' do
+    visit '/users'
+
+    expect(page).to have_content('John')
+    expect(page).to have_content('Alice')
+  end
```
Comment (Nemwel-Boniface) [inline spec/views/users/users_index_spec.rb:19]: - [ ] The error message indicates that there is an unknown attribute 'email' for the User model. This means that the User model in your Rails application does not have an 'email' attribute defined. If you check your schema you will see that indeed there is no "email" field for the user table.

image

File: spec/views/users/users_index_spec.rb [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+
+  it 'I can see the number of posts each user has written' do
+    visit '/users'
+    expect(page).to have_content('Number of posts: 2', count: 1)
+    expect(page).to have_content('Number of posts: 1', count: 1)
+  end
```
Comment (Nemwel-Boniface) [inline spec/views/users/users_index_spec.rb:25]: - [ ] This error is the same as the one above. KIndly refer to what I explained above.

image

File: spec/views/users/users_index_spec.rb [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+
+  it 'I can see the profile picture for each user' do
+    visit '/users'
+
+    expect(page).to have_selector("img[src='https://example.com/john.jpg']")
+    expect(page).to have_selector("img[src='https://example.com/alice.jpg']")
+  end
```
Comment (Nemwel-Boniface) [inline spec/views/users/users_index_spec.rb:32]: - [ ] This error is the same as the first comment above. KIndly refer to what I explained above.

image

File: spec/views/users/users_index_spec.rb [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+
+  it 'links to user show pages' do
+    visit '/users'
+
+    click_link 'John'
+
+    user_id = User.find_by(name: 'John').id
+    expect(current_path).to eq("/users/#{user_id}")
+  end
```
Comment (Nemwel-Boniface) [inline spec/views/users/users_index_spec.rb:41]: - [ ] This error is the same as the first comment above. KIndly refer to what I explained above.

image

File: spec/views/users/users_index_spec.rb [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+
+  it 'displays user information on show page' do
+    user = User.create(name: 'Jane', email: 'jane@example.com', photo: 'https://example.com/jane.jpg')
+
+    visit "/users/#{user.id}"
+
+    expect(page).to have_content('Jane')
+    expect(page).to have_selector("img[src='https://example.com/jane.jpg']")
+  end
+end
```
Comment (Nemwel-Boniface) [inline spec/views/users/users_index_spec.rb]: - [ ] This error is the same as the first comment above. KIndly refer to what I explained above.

image

**Review by Nemwel-Boniface — CHANGES_REQUESTED** [review]
## @HERMON-1995  :wave: 

### This is a manual review
# :recycle:  Request Changes :recycle: 

You did well adding the tests that were not included. However, the tests written are not passing and they seem to be failing. Have a look below at the specific tests that are failing.


- [ ] Kindly have a look at the comments left below. 
- [ ] KIndly ensure that the failing tests are passing.




Cheers and Happy coding!

File: spec/views/users/users_index_spec.rb (hunk 8 above)
Comment (HERMON-1995) [reply spec/views/users/users_index_spec.rb]: Ooops It looks like I have mistaken it with bio attribute at the last minute 😏, Sorry about that! Now all the tests are passing, Thank you! 😁

**Comment by bellom** [comment]
Hi @HERMON-1995,

### This is a manual review

# Approved 👍 

Cheers and Happy coding!:clap::clap::clap:

**Pull request merged**